### ISSUE TITLE
Cannot use cache.modify on relay fields after update to 3.2.1

### ISSUE THREAD

**Issue by anark** [apollographql/apollo-client#7088]
**Intended outcome:**
Prior to upgrading to 3.2.1 and using the relayStylePagination field policy the following would work when updating the cache after a mutation.  After update to 3.2.1 this no longer works due to the fact that the cache now stores edges inside wrapper objects.

```
  update: (cache, {data}) => {
     cache.modify({
         fields: {
             todos: (existingTodos) => {
                console.log(existingTodos, existingTodos.edges, readField('edges', existingTodos))
                const newTodoEdgeRef = cache.writeFragment({
                  data: data.newTodo?.newTodoEdge,
                  fragment: FRAGMENT,
                })
                return {
                  ...existingTodos,
                  edges: [...existingTodos.edges, newTodoEdgeRef],
                }
             }
         }
     }
  }
```

The initial change made(https://github.com/apollographql/apollo-client/pull/7023/commits/4746ac53c7a4eeeb893bbd6a0413cb33a524a8db) solved my original issue without adding this complexity.

**Actual outcome:**

This no longer works due to the fact that there is no 'edges' field on existingTodos.

**How to reproduce the issue:**

Set a field to use relayStylePagination field policy and try to use a cache.modify function to update the existing edges similar to the snippet above

**Versions**

3.2.1

**Comment by benjamn** [comment]
I introduced the wrapper objects in #7023 because I wanted to preserve the way `relayStylePagination` keeps track of edge cursors that come from `pageInfo`, given that `edges` might now contain `Reference` objects. Previously, we could just initialize the `edge.cursor` property of the first and last edges based on `pageInfo.has{Previous,Next}Page` (which is critical for continuous pagination, if you're not otherwise requesting `edges { cursor }`). Now that `edge` might be a `Reference`, it seemed awkward to try to update the `cursor` field in the normalized object referred to by the `Reference` during a `merge` function, so I introduced the wrapper layer to create a reliable place to store the `cursor` strings for pagination purposes.

That said, I don't love the extra complexity either, and I agree this was probably too big of a change for a patch version bump, given the very real possibility of `cache.modify` usage on `relayStylePagination` fields.

I wonder if there's something cleverer we can do here, to avoid changing the shape of the internal data. `Reference` objects are just objects, after all, so maybe we could tag them with extra properties?
```ts
type TEdge<TNode> = {
  cursor?: string;
  node: TNode;
} | (Reference & { cursor: string });
```
and then get rid of the `TEdgeWrapper` type?

**Comment by anark** [comment]
I think edges having ID's(and therefore being references) is quite rare and not worth the added complexity of the wrappers.

That said it is quite possible. Your suggestions of
```
type TEdge<TNode> = {
  cursor?: string;
  node: TNode;
} | (Reference & { cursor: string });
```

Sounds promising.

My only question is if edges should ever be a reference at all, as they will always be connecting two other references it may make sense not to store them as references

**Comment by benjamn** [comment]
@anark While I agree it might be convenient to assume edges are never normalized references, I also want the `relayStylePagination` code to be as robust as we can make it. I went ahead and implemented the idea I mentioned above (#7103), and it's published in `@apollo/client@3.3.0-beta.8`, if you want to give it a try before we release it in 3.2.2.

**Comment by anark** [comment]
Thanks @benjamn, I tried out 3.3.0-beta.8 and it seems to work great.

Thank you for the quick solution.
